homeassistant/components/heos/media_player.py

```diff
@@ -1,152 +1,262 @@
+        """Handle player attribute updated."""
+        from pyheos import const
+        if self._player.player_id != player_id:
+            return
+        if event == const.EVENT_PLAYER_NOW_PLAYING_PROGRESS:
+            self._media_position_updated_at = utcnow()
+        await self.async_update_ha_state(True)
+
+    async def async_added_to_hass(self):
+        """Device added to hass."""
+        from pyheos import const
+        # Update state when attributes of the player change
+        self._signals.append(self._player.heos.dispatcher.connect(
+            const.SIGNAL_PLAYER_EVENT, self._player_update))
+        # Update state when available players change
+        self._signals.append(self._player.heos.dispatcher.connect(
+            const.SIGNAL_CONTROLLER_EVENT, self._controller_event))
+        # Update state upon connect/disconnects
+        self._signals.append(self._player.heos.dispatcher.connect(
+            const.SIGNAL_HEOS_EVENT, self._heos_event))
+
+    async def async_clear_playlist(self):
+        """Clear players playlist."""
+        await self._player.clear_queue()
+
+    async def async_media_pause(self):
+        """Send pause command."""
+        await self._player.pause()
+
+    async def async_media_play(self):
+        """Send play command."""
+        await self._player.play()
+
+    async def async_media_previous_track(self):
+        """Send previous track command."""
+        await self._player.play_previous()
+
+    async def async_media_next_track(self):
+        """Send next track command."""
+        await self._player.play_next()
+
+    async def async_media_stop(self):
+        """Send stop command."""
+        await self._player.stop()
+
+    async def async_mute_volume(self, mute):
+        """Mute the volume."""
+        await self._player.set_mute(mute)
+
+    async def async_set_shuffle(self, shuffle):
+        """Enable/disable shuffle mode."""
+        await self._player.set_play_mode(self._player.repeat, shuffle)
 
-    def _update_state(self):
-        self.async_schedule_update_ha_state()
+    async def async_set_volume_level(self, volume):
+        """Set volume level, range 0..1."""
+        await self._player.set_volume(volume * 100)
 
     async def async_update(self):
-        """Update the player."""
-        self._player.request_update()
+        """Update supported features of the player."""
+        from pyheos import const
+        control_to_support = {
+            const.CONTROL_PLAY: SUPPORT_PLAY,
+            const.CONTROL_PAUSE: SUPPORT_PAUSE,
+            const.CONTROL_STOP: SUPPORT_STOP,
+            const.CONTROL_PLAY_PREVIOUS: SUPPORT_PREVIOUS_TRACK,
+            const.CONTROL_PLAY_NEXT: SUPPORT_NEXT_TRACK
+        }
+        controls = self._player.now_playing_media.supported_controls
+        current_support = [control_to_support[control] for control in controls]
+        self._supported_features = reduce(ior, current_support,
```

Comment:
@MartinHjelmare I confirmed with Paulus that updating `supported_flags` in this platform is still required to support of which playback controls are available.  This anti-pattern is still needed until it gets updated (which I plan to open an arch issue to address).

Reply:
Be sure to read all of home-assistant/architecture#1 before addressing this issue :)

Reply:
Yes - hence why I'm going to open a new issue...  Heos is a meta media player, and each of the sources has different supported controls.  I.e. a radio station has no next/previous or pause.  Pandora only has next.  Local music has next/previous, etc... I'd propose the attribute route (`can_play_next`).  Paulus seemed open to it hence why I'll open a new issue to revisit.  Thanks!